Complex fluid solver

abarret opened the pull request:
This adds in the complex fluid solver and also several examples.

I'll work on creating some tests for the code.

boyceg commented on examples/complex_fluids/ex2/InterpolationUtilities.h:
```diff
@@ -0,0 +1,80 @@
+#ifndef included_InterpolationUtilities
```
Could this go into IBTK?

It would be good to have something that lets us easily interpolate fields at arbitrary locations in the main library.

abarret replied:
It could. It would need to be cleaned up a bit. I can create another PR for that. It also didn't pan out well for flow past a confined cylinder, although I think that's just the nature of that specific problem.

drwells replied:
@abarret see #776.

boyceg commented on examples/complex_fluids/ex2/InterpolationUtilities.h:
```diff
@@ -0,0 +1,80 @@
+                                const std::vector<SAMRAI::solv::RobinBcCoefStrategy<NDIM>*>& bc_coefs,
+                                const double data_time,
+                                const int depth = 0);
+    static double weightFcn(const std::vector<double>&, const std::vector<double>&);
```
I think that the most prevalent convention is for `static` function names *not* to use camel case, but instead to be all lowercase with words separated by underscores.
```suggestion
    static double weight_fcn(const std::vector<double>&, const std::vector<double>&);
```

boyceg commented on examples/complex_fluids/ex2/InterpolationUtilities.h:
```diff
@@ -0,0 +1,80 @@
+private:
+    static double interpolate(const double& x, const std::vector<int>& xi, const std::vector<double>& yi);
+
+    static double interpolateInBoxes(const CellIndex<NDIM>& idx,
```
```suggestion
    static double interpolate_in_boxes(const CellIndex<NDIM>& idx,
```

boyceg commented on include/ibamr/AdvDiffComplexFluidConvectiveOperator.h:
```diff
@@ -0,0 +1,173 @@
+    return UNKNOWN;
+}
+/*!
+ * \brief Class AdvDiffComplexFluidConvectiveOperator is a concrete ConvectiveOperator that implements a convective
```
This is really an UCD for the conformation tensor or extra stress, right? I wonder if it would be both clearer and shorter to name this something like `CCUpperConvectiveOperator` (`CC` = cell-centered).

abarret replied:
Pretty much. The way it's written now, a relaxation function has to be registered.

The relaxation function could just be registered as a source term for the advection diffusion integrator, which would leave this as an UCD operator. That would make it more general. I'll do that, and move the UCD operator into the `adv_diff` folder.

abarret replied:
Actually, that's not completely true. This advects the conformation tensor _or_ the square root or logarithm of the tensor. So it's not just a UCD operator.

boyceg commented on include/ibamr/CFGiesekusRelaxation.h:
```diff
@@ -0,0 +1,134 @@
+ * \brief Class CFGiesekusRelaxation is a concrete CFRelaxationOperator that computes the relaxation function for the
+ * Giesekus fluid model.
+ */
+class CFGiesekusRelaxation : public IBAMR::CFRelaxationOperator
```
```suggestion
class CFGiesekusRelaxation : public CFRelaxationOperator
```

boyceg commented on include/ibamr/CFRelaxationOperator.h:
```diff
@@ -0,0 +1,131 @@
+namespace IBAMR
+{
+/*!
+ * \brief Class CFRelaxationOperator is an abstract class that provides an interface for specifying a relaxation
```
`CF` = "complex fluid"? Might be better to spell it out. Another alternative would be to add a `ComplexFluid` namespace within the `IBAMR` namespace. (@drwells, do you have an opinion?)

drwells replied:
I would prefer to use namespaces but this approach matches our current convention: e.g., we have `LSInitStrategy` instead of `LevelSet::InitStrategy`.

drwells replied:
i.e., lets keep the names this way for consistency.

boyceg commented on include/ibamr/CFGiesekusRelaxation.h:
```diff
@@ -0,0 +1,134 @@
+     *
+     * \note This constructor is not implemented and should not be used.
+     */
+    CFGiesekusRelaxation();
```
`= delete` ?

abarret replied:
Yes, and it should be made `public`, see https://stackoverflow.com/a/18931192

boyceg commented on include/ibamr/CFRoliePolyRelaxation.h:
```diff
@@ -0,0 +1,136 @@
+    CFRoliePolyRelaxation& operator=(const CFRoliePolyRelaxation& that);
+
+    double d_lambda_d, d_lambda_R, d_beta, d_delta;
+    bool d_sqr_root, d_log_conform;
```
Would this be clearer as something like an `enum`? I assume that the code requires at most one of these to be set to `true`.

boyceg replied:
Also, if the intent is for all of these implementations to work with the square root of the conformation tensor or the log conformation tensor, then this variable could go into the base class.

abarret replied:
If they are both `false`, the conformation tensor is evolved. I can create an `enum` for square root, logarithm, or regular.

boyceg commented on include/ibamr/ComplexFluidForcing.h:
```diff
@@ -0,0 +1,330 @@
+ *
+ * One can choose from the pre-programmed models Oldroyd-B, Giesekus, or Rolie-Poly. Optionally, you can also register a
+ * relaxation operator that computes the relaxation function. This class currently solves for the conformation tensor or
+ * the square root or logarithm of the conformation tensor. THe current assumption is that the stress is linearly
```
```suggestion
 * the square root or logarithm of the conformation tensor. The current assumption is that the stress is linearly
```

boyceg commented on include/ibamr/ComplexFluidForcing.h:
```diff
@@ -0,0 +1,330 @@
+ * One can choose from the pre-programmed models Oldroyd-B, Giesekus, or Rolie-Poly. Optionally, you can also register a
+ * relaxation operator that computes the relaxation function. This class currently solves for the conformation tensor or
+ * the square root or logarithm of the conformation tensor. THe current assumption is that the stress is linearly
+ * related to the conformation tensor through the elastic modulus.
```
Is it possible to relax this requirement in a subclass, or would it require more substantial modifications?

abarret replied:
The goal was to allow a function to be registered that does this transformation automatically. However, nobody was particularly interested in models that aren't linearly related, so I never got around to it. It should be a relatively painless change. There are a couple of `TODO:` comments in the code where this assumption is made.

boyceg commented on include/ibamr/ComplexFluidForcing.h:
```diff
@@ -0,0 +1,330 @@
+namespace IBAMR
+{
+/*!
+ * \brief Class ComplexFluidForcing provides an interface for specifying a viscoelastic stress to be added to the
```
Would consider `PolymericStressForcing` or `ExtraStressForcing`.

boyceg replied:
Also, for classes that require a particular data centering, we usually put the data centering in the class name. (Not sure if this is the best thing to do or not.)

abarret replied:
This works for both the cell centered and side centered fluid solver. The only difference is the stencil for the divergence of the conformation tensor.

boyceg replied:
But the conformation tensor is required to be cell centered?

abarret replied:
Yes, but only because the advection diffusion solver requires it.

I'm not sure how it should be named. All the INS forcing functions are named based on which solver they work with.

boyceg commented on include/ibamr/ComplexFluidForcing.h:
```diff
@@ -0,0 +1,330 @@
+                               bool initial_time,
+                               bool /*richardson_extrapolation_too*/);
+
+    static void applyGradientDetectorCallback(SAMRAI::tbox::Pointer<SAMRAI::hier::BasePatchHierarchy<NDIM>> hierarchy,
```
Possibly:
```suggestion
    static void apply_gradient_detector_callback(SAMRAI::tbox::Pointer<SAMRAI::hier::BasePatchHierarchy<NDIM>> hierarchy,
```

boyceg commented on include/ibamr/ComplexFluidForcing.h:
```diff
@@ -0,0 +1,330 @@
+    /*!
+     * \brief Get values from input file
+     *
+     * \param
```
```suggestion
```

boyceg commented on src/complex_fluids/AdvDiffComplexFluidConvectiveOperator.cpp:
```diff
@@ -0,0 +1,464 @@
+    d_convec_oper->setAdvectionVelocity(d_u_idx);
+    d_convec_oper->setSolutionTime(d_solution_time);
+    Pointer<CartesianGridGeometry<NDIM>> grid_geom = d_hierarchy->getGridGeometry();
+    // Copy data to side centered velocity field
```
We should probably have a helper function to do this kind of thing in IBTK (e.g. in `PatchMathOps`).

boyceg commented on src/complex_fluids/CFGiesekusRelaxation.cpp:
```diff
@@ -0,0 +1,153 @@
+        const CellIndex<NDIM>& idx = i();
+#if (NDIM == 2)
+        double Q_xx, Q_yy, Q_xy;
+        if (d_sqr_root)
```
Is the following code common to all formulations that use sqrt or log conformation tensor? _Might_ be worth putting it as a helper function in the base class.

boyceg commented on src/complex_fluids/CFGiesekusRelaxation.cpp:
```diff
@@ -0,0 +1,153 @@
+            Q_xz = (*in_data)(idx, 4);
+            Q_xy = (*in_data)(idx, 5);
+        }
+        (*ret_data)(idx, 0) =
```
This last bit looks like the only part that is specific to Giesekus. I wonder if it would be better to template the base class on the function that evaluates these kinds of expressions, or to have users pass in these kinds of functions as simple C functions (similar to the stress and force functions in `IBFEMethod`). Presumably would make it easier for users to specify their own models.

boyceg commented on src/complex_fluids/ComplexFluidForcing.cpp:
```diff
@@ -0,0 +1,957 @@
+        d_conc_bc_coefs.resize(NDIM * (NDIM + 1) / 2);
+        for (int d = 0; d < NDIM * (NDIM + 1) / 2; ++d)
+        {
+            ostringstream conc_bc_name;
```
I think you can just use `std::string` in C++11?

boyceg commented on src/complex_fluids/ComplexFluidForcing.cpp:
```diff
@@ -0,0 +1,957 @@
+    d_adv_diff_integrator->setConvectiveOperator(d_W_cc_var, d_convec_oper);
+
+    // Register relaxation function
+    if (d_fluid_model == "OLDROYDB")
```
Would be nice to make this more extensible (e.g. to allow for user specified models).

boyceg commented on src/complex_fluids/ComplexFluidForcing.cpp:
```diff
@@ -0,0 +1,957 @@
+    if (d_log_det)
+    {
+        d_max_det = 0.0;
+        d_min_det = 1.0e10;
```
`std::numeric_limits<double>::max()` ?

abarret replied:
This is something I'm not sure about. Will this affect `max` and `min` MPI reductions?

boyceg replied:
MPI reductions will be fine.

Practically speaking, 1e10 is also probably fine, but I think it is clearer to use `numeric_limits` for large values.

boyceg commented on src/complex_fluids/ComplexFluidForcing.cpp:
```diff
@@ -0,0 +1,957 @@
+} // exponentiateMatrix
+
+void
+ComplexFluidForcing::projectMatrix(const int data_idx,
```
Possibly:
```suggestion
ComplexFluidForcing::projectTensor(const int data_idx,
```
?

boyceg commented on src/complex_fluids/ComplexFluidForcing.cpp:
```diff
@@ -0,0 +1,957 @@
+            {
+                CellIndex<NDIM> i = *it;
+#if (NDIM == 2)
+                Eigen::Matrix2d c;
```
Might be useful to have a helper function that will construct the tensor given the patch data and a cell index?

drwells commented:
Its not a big deal, but (to follow up on the namespace question) I would prefer the file names to be, e.g., `CFForcing` to match the same naming convention we have with level set.

abarret commented:
I addressed all of the comments. I renamed things based on @drwells suggestion to be consistent with the current naming conventions, e.g. `CFINSForcing` and `CFRelaxationOperator`.

Users can register relaxation operators which is an extension of `CartGridFunction`. The new class `CFRelaxationOperator` is needed to ensure that the correct patch data is used to compute the operator.

There are still no tests, I will work on that tomorrow.

boyceg commented on include/ibamr/CFUpperConvectiveOperator.h:
```diff
@@ -0,0 +1,148 @@
+/////////////////////////////// INCLUDES /////////////////////////////////////
+#include "IBAMR_config.h"
+
+#include "ibamr/AdvDiffCUIConvectiveOperator.h"
```
Looks like we need to run IWYU. :-)

abarret replied:
Oh right good catch. I forgot to delete these. I was testing different ways to allocate the convective operator. I'll double check the includes.

boyceg commented on src/complex_fluids/CFINSForcing.cpp:
```diff
@@ -0,0 +1,952 @@
+{
+    if (initial_time)
+    {
+        // TODO: The integrators do not call setDataOnPatchHierarchy for the initial iteration. This is a workaround for
```
@abarret can you create an issue outlining these issues?

abarret replied:
Yes.

boyceg commented:
Just following up on this. @drwells may have comments, but I think this is basically good to go as soon as there are tests.

abarret commented:
Added some tests. This is ready for a final review.

abarret commented:
Nevermind. These tests fail on master.

abarret commented:
Fixed the tests. Apparently another line of output has been added since I created the output. Now this is ready for a final review.

drwells commented on examples/complex_fluids/ex0/Makefile.am:
```diff
@@ -0,0 +1,30 @@
+
+##Dimension-dependent testers
+EXAMPLE_DRIVER = ex_main.cpp
+GTEST_DRIVER   = test_main.cpp
```
All the gtest stuff is gone now, so please get rid of this and the other gtest targets/variables.

abarret replied:
These were the merge conflicts. They are now gone

drwells commented on examples/complex_fluids/ex1/Makefile.am:
```diff
@@ -0,0 +1,30 @@
+GTEST_DRIVER   = test_main.cpp
+EXTRA_DIST     = input2d
+
+GTESTS   =
```
same

drwells commented on examples/complex_fluids/ex2/Makefile.am:
```diff
@@ -0,0 +1,39 @@
+EXTRA_DIST     = input2d
+
+
+GTESTS   =
```
same

drwells commented on examples/complex_fluids/ex2/ex_main.cpp:
```diff
@@ -0,0 +1,8 @@
+{
+    run_example(argc, argv);
+    return 0;
+}
```
These need to be merged back with `example.cpp`.

drwells commented on examples/complex_fluids/ex3/Makefile.am:
```diff
@@ -0,0 +1,30 @@
+
+##Dimension-dependent testers
+EXAMPLE_DRIVER = ex_main.cpp
+GTEST_DRIVER   = test_main.cpp
```
etc.

drwells commented on include/ibamr/CFOldroydBRelaxation.h:
```diff
@@ -0,0 +1,126 @@
+
+#include "ibamr/CFRelaxationOperator.h"
+#include "ibamr/ibamr_enums.h"
+#include "ibamr/namespaces.h" // IWYU pragma: keep
```
This shouldn't be in a header.

drwells commented on include/ibamr/CFRelaxationOperator.h:
```diff
@@ -0,0 +1,131 @@
+
+#include "ibamr/AdvDiffHierarchyIntegrator.h"
+#include "ibamr/ibamr_enums.h"
+#include "ibamr/namespaces.h" // IWYU pragma: keep
```
this shouldn't be in a header.

drwells commented on include/ibamr/CFRelaxationOperator.h:
```diff
@@ -0,0 +1,131 @@
+#include "ibamr/AdvDiffHierarchyIntegrator.h"
+#include "ibamr/ibamr_enums.h"
+#include "ibamr/namespaces.h" // IWYU pragma: keep
+#include <ibamr/app_namespaces.h>
```
this shouldn't be in a header.

drwells commented on include/ibamr/CFRoliePolyRelaxation.h:
```diff
@@ -0,0 +1,128 @@
+
+#include "ibamr/CFRelaxationOperator.h"
+#include "ibamr/ibamr_enums.h"
+#include "ibamr/namespaces.h" // IWYU pragma: keep
```
this shouldn't be in a header.

drwells commented on include/ibamr/CFUpperConvectiveOperator.h:
```diff
@@ -0,0 +1,144 @@
+#include "ibamr/StaggeredStokesPhysicalBoundaryHelper.h"
+#include "ibamr/ibamr_enums.h"
+#include "ibamr/ibamr_utilities.h"
+#include "ibamr/namespaces.h" // IWYU pragma: keep
```
this shouldn't be in a header.

drwells commented on include/ibamr/CFGiesekusRelaxation.h:
```diff
@@ -0,0 +1,126 @@
+
+#include "ibamr/CFRelaxationOperator.h"
+#include "ibamr/ibamr_enums.h"
+#include "ibamr/namespaces.h" // IWYU pragma: keep
```
this shouldn't be in a header.

drwells commented on include/ibamr/CFINSForcing.h:
```diff
@@ -0,0 +1,318 @@
+#include "ibamr/CFUpperConvectiveOperator.h"
+#include "ibamr/INSStaggeredHierarchyIntegrator.h"
+#include "ibamr/ibamr_enums.h"
+#include "ibamr/namespaces.h" // IWYU pragma: keep
```
this shouldn't be in a header.

drwells commented on tests/Makefile.am:
```diff
@@ -12,8 +12,8 @@
 include $(top_srcdir)/config/Make-rules
-TEST_DIRECTORIES = adv_diff advect IB IBTK IBFE multiphase_flow navier_stokes \
-physical_boundary refine spread wave_tank
+TEST_DIRECTORIES = adv_diff advect complex_fluidsIB IBTK IBFE multiphase_flow \
```
This should be `complex_fluids`

drwells commented on tests/Makefile.in:
```diff
@@ -405,8 +405,8 @@ IBAMR2d_LIBS = ${top_builddir}/lib/libIBAMR2d.a ${top_builddir}/ibtk/lib/libIBTK
 SUFFIXES = .f.m4
-TEST_DIRECTORIES = adv_diff advect IB IBTK IBFE multiphase_flow navier_stokes \
-physical_boundary refine spread wave_tank
+TEST_DIRECTORIES = adv_diff advect complex_fluidsIB IBTK IBFE multiphase_flow \
```
same (i.e., rerun `autoreconf`)

drwells commented:
23 new tests :tada:

drwells commented:
There is something wrong with the tests. I get `nan` and `-nan` for the L1 and L2 norms in 2D and huge values (circa 25) in 3D. For example:
```
[drwells@archway complex_fluids]$ ./cf_forcing_op_01_2d ./cf_forcing_op_01_2d.side.input
[drwells@archway complex_fluids]$ cat output 
AdvDiffHierarchyIntegrator::initializePatchHierarchy(): tag_buffer = 0
Conformation tensor is SPD
Error norms:
  L1-norm:  nan
  L2-norm:  -nan
  max-norm: 0.010088326
```
and
```
[drwells@archway complex_fluids]$ ./cf_forcing_op_01_3d ./cf_forcing_op_01_3d.side.input
[drwells@archway complex_fluids]$ cat output 
AdvDiffHierarchyIntegrator::initializePatchHierarchy(): tag_buffer = 0
Conformation tensor is SPD
Error norms:
  L1-norm:  25.848664
  L2-norm:  28.791885
  max-norm: 68.751187
```

@abarret any idea why this might be happening?

drwells commented:
In particular, the following tests fail for me:
```
    complex_fluids/cf_four_roll_mill.logarithm.input
    complex_fluids/cf_forcing_op_01_2d.cell.input
    complex_fluids/cf_four_roll_mill.square_root.input
    complex_fluids/cf_four_roll_mill.standard.input
    complex_fluids/cf_forcing_op_01_3d.side.input
    complex_fluids/cf_forcing_op_01_2d.side.input
    complex_fluids/cf_forcing_op_01_3d.cell.input
```

drwells commented:
I get similar results for `cf_four_roll_mill`: some values seem to be set to the maximum double precision value.

abarret commented:
I'm not sure. All the tests pass for me on both a debugging and optimized build.

abarret commented:
@drwells I just pushed another commit that fixes some namespacing issues in the tests as well as fixes the Makefile. I wouldn't expect this to fix the tests though.

drwells commented:
I'm recompiling in debug mode - it might give a hint as to why things don't work.

drwells commented:
Hm - I get the same problems in debug mode too.

abarret commented:
What happens if you run `examples/complex_fluids/ex0`? It's the same setup as `cf_four_roll_mill`. You'll have to visualize the results since the example doesn't print solution norms.

drwells commented:
ex0 does print error norms. They are all pretty small:
```
Error in u at time 4:
  L1-norm:  0.0001324564936
  L2-norm:  8.178764174e-05
  max-norm: 0.0001021703581
Error in sxx at time 4:
  L1-norm:  2.943169048e-05
  L2-norm:  2.913995926e-05
  max-norm: 0.0002343565716
Error in syy at time 4:
  L1-norm:  2.101223228e-05
  L2-norm:  2.54879698e-05
  max-norm: 0.0001644535124
Error in sxy at time 4:
  L1-norm:  0.000100835169
  L2-norm:  8.211012716e-05
  max-norm: 0.0003908311637
```

abarret commented:
Sorry, I switched the order of the examples and keep forgetting about it. The test is the same as `examples/complex_fluids/ex1`.

drwells commented:
I have the same problem with that example - its even in the log file.

```
Conformation tensor is SPD
Largest det:  0
Smallest det: 1.79769e+308
```

drwells commented:
I don't have visit installed (I'm working on it) but I can see that, at timestep 10, the conformation tensor is all `-nan`s.

drwells commented:
I can see the same problem after a single time step as well.

drwells commented:
The tests pass correctly with my standard test setup on my workstation - the problem is on my desktop computer.

abarret commented:
This is a long shot, but could it be an issue with Eigen? I use Eigen pretty heavily to convert between different forms of the stress tensor. Can you run `examples/IB/explicit/ex4` without issues?

drwells commented:
Thats a good idea - I am using a much newer version of eigen than the bundled one. Let me take a look.

drwells commented:
ex4 doesn't work right now since I broke IBFESurfaceMethod - see #775.

abarret commented:
`IB/explicit/ex4` also uses Eigen pretty heavily (to compute matrix square roots). Note it's _not_ the IBFE example.

drwells commented:
oh, whoops - I was looking at the complex fluids example 4.

drwells commented:
I don't see any weird values in `IB/explicit/ex4` either.

abarret commented:
Well I'm stumped. I can't think of anything different going on in `CFINSForcing` other than using Eigen to compute matrix exponentials and eigenvalues.

drwells commented:
I'm not sure either. I'm recompiling IBAMR with the bundled version of eigen to check.

drwells commented:
Nope: changing the version of eigen doesn't help. That's probably a good thing...

drwells commented:
@abarret would you please add, to `cp_forcing_op_01.cpp`, the header
```cpp
#include <fenv.h>
```
and, to `main()`
```cpp
feenableexcept(FE_ALL_EXCEPT);
```
and see if it makes any difference in execution?

drwells commented:
For some horrifying reason this is actually tripped inside PETSc's initialization. Could you try
```cpp
feenableexcept(FE_INVALID);
```
instead?

abarret commented:
> For some horrifying reason this is actually tripped inside PETSc's initialization. Could you try
> 
> ```c++
> feenableexcept(FE_INVALID);
> ```
> 
> instead?

This didn't change anything for me.

drwells commented:
This looks like a bug in Eigen. I have no idea how this code works for you. Consider the code circa line 680 in `SelfAdjointEigenSolver.h`:
```cpp
    Scalar scale = mat.cwiseAbs().maxCoeff();
    scale = (std::max)(scale,Scalar(1));
    MatrixType scaledMat = mat / scale;
    
    // Compute the eigenvalues
    computeRoots(scaledMat,eivals);
    
    // compute the eigen vectors
    if(computeEigenvectors)
    {
      scaledMat.diagonal().array () -= eivals(1);
      Scalar a2 = numext::abs2(scaledMat(0,0));
      Scalar c2 = numext::abs2(scaledMat(1,1));
      Scalar b2 = numext::abs2(scaledMat(1,0));
      if(a2>c2)
      {
        eivecs.col(1) << -scaledMat(1,0), scaledMat(0,0);
        eivecs.col(1) /= sqrt(a2+b2);
      }
      else
      {
        eivecs.col(1) << -scaledMat(1,1), scaledMat(1,0);
        eivecs.col(1) /= sqrt(c2+b2);
      }

      eivecs.col(0) << eivecs.col(1).unitOrthogonal();
    }
```

here `mat` is `2.707, 0, 0, 2.707`: a nice diagonal matrix. Naturally `scale = 2.707`. As the eigenvalues of `scaledMat` are, unsurprisingly, `1`, and `1`,  `a2 == c2 == b2 == 0.0`: i.e., this code will always be broken for a scaled identity matrix since we divide something by `sqrt(c2 + b2)`.

drwells commented:
Could you try poking around with setting a breakpoint in that function inside eigen and see, when `mat` is a scaled identity matrix, whether or not things work?

abarret commented:
Well I know why it doesn't crash. My copy of Eigen checks if the eigenvalues are the same.

abarret commented:
I'm using the bundled copy of Eigen from libMesh 1.2.1

abarret commented:
Here's that function in my copy of `SelfAdjointEigenSolver.h`

    Scalar scale = mat.cwiseAbs().maxCoeff();
    scale = (std::max)(scale,Scalar(1));
    MatrixType scaledMat = mat / scale;

    // Compute the eigenvalues
    computeRoots(scaledMat,eivals);

    // compute the eigen vectors
    if(computeEigenvectors)
    {
      if((eivals(1)-eivals(0))<=abs(eivals(1))*Eigen::NumTraits<Scalar>::epsilon())
      {
        eivecs.setIdentity();
      }
      else
      {
        scaledMat.diagonal().array () -= eivals(1);
        Scalar a2 = numext::abs2(scaledMat(0,0));
        Scalar c2 = numext::abs2(scaledMat(1,1));
        Scalar b2 = numext::abs2(scaledMat(1,0));
        if(a2>c2)
        {
          eivecs.col(1) << -scaledMat(1,0), scaledMat(0,0);
          eivecs.col(1) /= sqrt(a2+b2);
        }
        else
        {
          eivecs.col(1) << -scaledMat(1,1), scaledMat(1,0);
          eivecs.col(1) /= sqrt(c2+b2);
        }

        eivecs.col(0) << eivecs.col(1).unitOrthogonal();
      }
    }

drwells commented:
Interesting - I'll check again in a moment, but I am pretty sure that this means that it is a bug with the bundled copy of eigen. Let me recompile again, this time triple-checking that I pick up my system copy.

drwells commented:
I can also only get the test code
```cpp
#include <Eigen/Eigenvalues>

#include <iostream>

int main()
{

  Eigen::Matrix2d c;
  Eigen::SelfAdjointEigenSolver<Eigen::Matrix2d> eigs;
  c(0, 0) = 2.707;
  c(0, 1) = 0.0;
  c(1, 0) = 0.0;
  c(1, 1) = 2.707;
  eigs.computeDirect(c);
  Eigen::Matrix2d eig_vals(Eigen::Matrix2d::Zero());
  eig_vals(0, 0) = eigs.eigenvalues()(0);
  eig_vals(1, 1) = eigs.eigenvalues()(1);
  eig_vals(0, 0) = std::max(eig_vals(0, 0), 0.0);
  eig_vals(1, 1) = std::max(eig_vals(1, 1), 0.0);
  Eigen::Matrix2d eig_vecs = eigs.eigenvectors();

  std::cout << eig_vals(0, 0) << std::endl;
  std::cout << eig_vals(1, 1) << std::endl;
  std::cout << eig_vecs(0, 0) << std::endl;
  std::cout << eig_vecs(1, 0) << std::endl;
}
```
to output NaNs if I use our bundled copy of eigen, which strongly suggests that that is the problem.

abarret commented:
This is a bug that was fixed after Eigen 3.2
https://gitlab.com/libeigen/eigen/issues/1013

drwells commented:
Let me bump #754 to the milestone to fix this.

drwells commented:
Note that even though the gitlab issue is new this bug was fixed years ago:
https://gitlab.com/libeigen/eigen/commit/847bb317cdb061a07f170e6373375eb2d302846d

abarret commented:
Yeah. They switched from bitbucket to gitlab yesterday.

drwells commented:
Fixing Eigen fixes 6 of the 7 failing tests. I still get different results for `cf_forcing_op_01_2d.cell.input`:
```
[drwells@archway complex_fluids]$ ./cf_forcing_op_01_2d ./cf_forcing_op_01_2d.cell.input 
[drwells@archway complex_fluids]$ cat output 
AdvDiffHierarchyIntegrator::initializePatchHierarchy(): tag_buffer = 0
Conformation tensor is SPD
Error norms:
  L1-norm:  3.2289116
  L2-norm:  2.2831853
  max-norm: 1.6144558
[drwells@archway complex_fluids]$ cat cf_forcing_op_01_2d.cell.output 
AdvDiffHierarchyIntegrator::initializePatchHierarchy(): tag_buffer = 0
Conformation tensor is SPD
Error norms:
  L1-norm:  0.051387699
  L2-norm:  0.040295003
  max-norm: 0.040100971
```
Its weird that the side-centered test passes and the cell-centered one fails.

drwells commented:
Its also weird that the 3D version passes

abarret commented:
What happens if you include the line
`   project_conformation_tensor = FALSE`
in the `ComplexFluid` database of the input file `cf_forcing_op_01_2d.cell.input`?

That should skip the calculation of the eigenvalues/vectors. It should also have no effect since the conformation tensor is already SPD.

drwells commented:
That doesn't have any effect on the output file.

drwells commented:
I'm dumb - I changed the input file to `N = 4` while I was debugging and forgot to change it back.

drwells commented on examples/complex_fluids/ex4/example.cpp:
```diff
@@ -0,0 +1,573 @@
+            Elem* const elem = *el;
+            for (unsigned int side = 0; side < elem->n_sides(); ++side)
+            {
+                const bool at_mesh_bdry = !elem->neighbor(side);
```
```suggestion
                const bool at_mesh_bdry = !elem->neighbor_ptr(side);
```

drwells commented on examples/complex_fluids/ex4/example.cpp:
```diff
@@ -0,0 +1,573 @@
+                for (unsigned int k = 0; k < elem->n_nodes(); ++k)
+                {
+                    if (!elem->is_node_on_side(k, side)) continue;
+                    Node& n = *elem->get_node(k);
```
```suggestion
                    Node& n = *elem->node_ptr(k);
```

drwells commented on examples/complex_fluids/ex2/convergence_tester.cpp:
```diff
@@ -0,0 +1,563 @@
+            {
+                Pointer<Patch<NDIM> > patch = level->getPatch(p());
+                const Box<NDIM>& box = patch->getBox();
+                const Index<NDIM>& idx_low = box.lower();
```
```suggestion
                const hier::Index<NDIM>& idx_low = box.lower();
```

drwells commented:
This should be ready once we fix the compilation issues (these matter with Eigen 3.3 and newer versions of libMesh).

abarret commented:
Fixed the suggestions and squashed into two commits. Should we wait to merge until #779 is merged?

drwells commented:
@abarret now that you are an admin you could merge it :)
